core/src/data_sources/splitter.rs

```diff
@@ -79,37 +81,111 @@ impl Splitter for BaseV0Splitter {
         let re = Regex::new(r"\s+").unwrap();
         let clean = re.replace_all(text, " ");
         let clean = clean.trim();
-
         // Get the embedder and initialize.
         let mut embedder = provider(provider_id).embedder(model_id.to_string());
         embedder.initialize(credentials).await?;
 
         // Encode the clean text.
         let encoded = embedder.encode(clean).await?;
+        let mut decoded = vec![];
+
+        let mut start = 0;
+        while start < encoded.len() {
+            let end = cmp::min(start + max_chunk_size - 1, encoded.len() - 1);
 
-        // Split the encoded text into chunks of size max_chunk_size.
-        let mut chunks = Vec::new();
-        let mut chunk = Vec::new();
-        let mut chunk_size = 0;
-        for token in encoded {
-            chunk.push(token);
-            chunk_size += 1;
-            if chunk_size >= max_chunk_size {
-                chunks.push(chunk);
-                chunk = Vec::new();
-                chunk_size = 0;
+            match decode_chunk_with_remainder(&embedder, &encoded[start..=end]).await {
+                Ok((chunk_decoded, chunk_remainder)) => {
+                    
+                    if chunk_remainder.is_some() {
+                        let chunk_remainder_value = chunk_remainder.unwrap();
+                        if chunk_remainder_value.len() == max_chunk_size {
+                            return Err(anyhow!("The remainder chunk is the same size as the max chunk size - can't keep decoding. max_chunk_size: {}", max_chunk_size));
```

Comment:
let's enforce the 100 char wrap. You can do: 
```
"b laba lalbal \
  aslkd sadl"
```

The error message will buble up all the way to the user. So we should probably use something more generic, like "failed to chunk document"?

Reply:
(esp. as this error should never be hit)

Reply:
If we enforce `max_chunk_size > 4` of course